datafusion/physical-plan/src/projection.rs

```diff
@@ -1459,4 +1633,608 @@ mod tests {
+            },
+        ]);
+
+        // Merge should produce: SELECT c@2 AS col1, b@1 AS col2
+        let merged = base_projection.try_merge(&top_projection)?;
+        assert_eq!(merged.as_ref().len(), 2);
+
+        // Check first expression (col1 should reference c@2)
+        let col1_expr = merged.as_ref()[0].expr.as_any().downcast_ref::<Column>();
+        assert!(col1_expr.is_some());
+        let col1 = col1_expr.unwrap();
+        assert_eq!(col1.name(), "c");
+        assert_eq!(col1.index(), 2);
+        assert_eq!(merged.as_ref()[0].alias, "col1");
+
+        // Check second expression (col2 should reference b@1)
+        let col2_expr = merged.as_ref()[1].expr.as_any().downcast_ref::<Column>();
+        assert!(col2_expr.is_some());
+        let col2 = col2_expr.unwrap();
+        assert_eq!(col2.name(), "b");
+        assert_eq!(col2.index(), 1);
+        assert_eq!(merged.as_ref()[1].alias, "col2");
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_merge_with_expressions() -> Result<()> {
+        // First projection: SELECT c@2 AS x, b@1 AS y, a@0 AS z
+        let base_projection = Projection::new(vec![
+            ProjectionExpr {
+                expr: Arc::new(Column::new("c", 2)),
+                alias: "x".to_string(),
+            },
+            ProjectionExpr {
+                expr: Arc::new(Column::new("b", 1)),
+                alias: "y".to_string(),
+            },
+            ProjectionExpr {
+                expr: Arc::new(Column::new("a", 0)),
+                alias: "z".to_string(),
+            },
+        ]);
+
+        // Second projection: SELECT x@0 + 1 AS c1, y@1 + z@2 AS c2
+        let top_projection = Projection::new(vec![
+            ProjectionExpr {
+                expr: Arc::new(BinaryExpr::new(
+                    Arc::new(Column::new("x", 0)),
+                    Operator::Plus,
+                    Arc::new(Literal::new(ScalarValue::Int32(Some(1)))),
+                )),
+                alias: "c1".to_string(),
+            },
+            ProjectionExpr {
+                expr: Arc::new(BinaryExpr::new(
+                    Arc::new(Column::new("y", 1)),
+                    Operator::Plus,
+                    Arc::new(Column::new("z", 2)),
+                )),
+                alias: "c2".to_string(),
+            },
+        ]);
+
+        // Merge should produce: SELECT c@2 + 1 AS c1, b@1 + a@0 AS c2
+        let merged = base_projection.try_merge(&top_projection)?;
+        assert_eq!(merged.as_ref().len(), 2);
+        assert_eq!(merged.as_ref()[0].alias, "c1");
+        assert_eq!(merged.as_ref()[1].alias, "c2");
+
+        // Check that the expressions are BinaryExpr (not just Column)
+        assert!(merged.as_ref()[0].expr.as_any().is::<BinaryExpr>());
+        assert!(merged.as_ref()[1].expr.as_any().is::<BinaryExpr>());
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_merge_docstring_example() -> Result<()> {
+        // Example from the docstring:
```

Comment:
How about you just make this a docstring test? That would ensure it is run as well as improve the docs